chart/kubeapps/templates/kubeapps-frontend-config.yaml

```diff
@@ -58,10 +58,7 @@ data:
       # used with an auth-proxy where k8s credentials never leave
       # the cluster). See clusters option.
       {{- range .Values.clusters }}
-      location ~* /api/clusters/{{ .name }} {
-        rewrite /api/clusters/{{ .name }}/(.*) /$1 break;
-        rewrite /api/clusters/{{ .name }} / break;
```

Comment:
@absoludity @antgamdia unfortunately, this change doesn't seem to work as expected.

```
172.17.0.24 - - [05/Feb/2021:22:59:20 +0000] "GET /api/clusters/default/ HTTP/1.1" 502 552 "http://apps.coreweave.test:31035/" "Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/88.0.4324.146 Safari/537.36"
2021/02/05 22:59:20 [error] 27#27: *5 no resolver defined to resolve gatekeeper-kubernetes-ingress.default, client: 172.17.0.24, server: _, request: "GET /api/clusters/default/ HTTP/1.1", host: "apps.coreweave.test:31035", referrer: "http://apps.coreweave.test:31035/"
```
Either of you guys see what might be wrong?

Reading the existing config, it just seems to remove `/api/clusters/{{ .name }}` from the URL before proxying the request to `.apiServiceUrl`.

My thinking was to just place the "extra" group (i.e. what comes after `/api/clusters/default`)  into the location itself so I can then `proxy_pass http://some-host/k8s/$1` and thereby avoid [this NGINX error](https://stackoverflow.com/questions/53353572/proxy-pass-cannot-have-uri-part-in-location-given-by-regular-expression).

Reply:
> Reading the existing config, it just seems to remove `/api/clusters/{{ .name }}` from the URL before proxying the request to `.apiServiceUrl`.

That's correct - the kubeapps uri defines the cluster name and we use that to proxy the request to the correct api service url with the remaining path.

> 
> My thinking was to just place the "extra" group (i.e. what comes after `/api/clusters/default`) into the location itself so I can then `proxy_pass http://some-host/k8s/$1` and thereby avoid [this NGINX error](https://stackoverflow.com/questions/53353572/proxy-pass-cannot-have-uri-part-in-location-given-by-regular-expression).

I don't understand what needs changing here, given that I think you want the following (I think?):
```
http://kubeapps.some-host/k8s/api/v1/namespaces -> http://gateway.some-host/k8s/api/v1/namespaces
```

which wouldn't require any location/rewrite changes from the above, just the proxy pass. But there must be something I'm missing. Can you let us know exact examples of what you're trying to achieve right here (sorry if you've already said this and I've lost it in the other threads/PRs).

It looks like what you have here is already a workaround for some other issue, I'd like to see the original thing you tried, know why you're trying it and what the issue is, before looking at possible workarounds :)

Reply:
@absoludity sure thing, here is what I am currently trying to do.

### NGINX configuration for k8s API location proxy

I have `.apiServiceURI` set to `http://gatekeeper-kubernetes-ingress.default/k8s/` - this way I can use the `/k8s` portion of the URI in my API gateway so that it proxies ingress to the k8s API backend (and I don't have to set some special `proxy_set_header Host`).

```
      location ~* /api/clusters/default {
        rewrite /api/clusters/default/(.*) /$1 break;
        rewrite /api/clusters/default / break;

      # Helm returns a nil pointer error when accessing foo.bar if foo doesn't
      # exist, even with the `default` function.
      # See https://github.com/helm/helm/issues/8026#issuecomment-756538254
        # Otherwise we route directly through to the clusters with existing credentials.
        proxy_pass http://gatekeeper-kubernetes-ingress.default/k8s/;
        include "./server_blocks/k8s-api-proxy.conf";
      }
```

However, this is not a valid NGINX configuration:

```
(⎈ minikube:default)➜  gatekeeper git:(mc/cloud-charts) ✗ k logs -f gatekeeper-kubeapps-67c86c5574-g24sr
nginx 17:00:25.14
nginx 17:00:25.14 Welcome to the Bitnami nginx container
nginx 17:00:25.15 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-nginx
nginx 17:00:25.15 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-nginx/issues
nginx 17:00:25.15
nginx 17:00:25.15 INFO  ==> ** Starting NGINX setup **
nginx 17:00:25.24 INFO  ==> Validating settings in NGINX_* env vars
nginx 17:00:25.25 INFO  ==> Initializing NGINX

nginx 17:00:25.26 INFO  ==> ** NGINX setup finished! **
nginx 17:00:25.35 INFO  ==> ** Starting NGINX **
2021/02/12 17:00:25 [emerg] 1#1: "proxy_pass" cannot have URI part in location given by regular expression, or inside named location, or inside "if" statement, or inside "limit_except" block in /opt/bitnami/nginx/conf/server_blocks/vhost.conf:33
nginx: [emerg] "proxy_pass" cannot have URI part in location given by regular expression, or inside named location, or inside "if" statement, or inside "limit_except" block in /opt/bitnami/nginx/conf/server_blocks/vhost.conf:33
```

Reply:
That makes sense, but it's what I already understood. The bit that's not clear to me is why what you've stated you need requires a change to the location. Let me use some examples:

Is this an example of a URI you want to proxy here:
```
https://my-kubeapps-domain.com/api/clusters/cluster-name/k8s/api/v1/namespaces
```
?

If so, the existing:

```
      location ~* /api/clusters/{{ .name }} {
```

should match the URL, and the existing:

```
        rewrite /api/clusters/{{ .name }}/(.*) /$1 break;
        rewrite /api/clusters/{{ .name }} / break;
```

should ensure that the *path* of the request is rewritten from `/api/clusters/cluster-name/k8s/api/v1/namespaces` to `/k8s/api/v1/namespaces` . Which is, afaict, exactly what you want? So I'm guessing something didn't quite work when you tried that and hence you're now trying to use an explicit regex is the location etc.? But I'm not sure - that's what I'm trying to understand... why the default config for the location+rewrite didn't do what you want.

On the other hand, if what you're trying to do is have an incoming URL *without* the `/k8s/` such as:

```
https://my-kubeapps-domain.com/api/clusters/cluster-name/api/v1/namespaces
```

and want that routed to your `http://gatekeeper-kubernetes-ingress.default` with the extra `/k8s/`, ie. so the path would be rewritten to `/k8s/api/v/namespaces`, then it should just require a small update to the rewrite rules:

```
        rewrite /api/clusters/{{ .name }}/(.*) /k8s/$1 break;
        rewrite /api/clusters/{{ .name }} /k8s/ break;
``` 

It's just not clear to me which of those (or something else) you're trying to do. If that works, then yes, we'll need to find some way that (1) allows you to include the `k8s` path in the `.apiServerURL` while stripping it off for the `proxy_path` directive (should be easy), and (2) possibly be able to include a subpath in the rewrite.

Reply:
> Is this an example of a URI you want to proxy here:
> 
> ```
> https://my-kubeapps-domain.com/api/clusters/cluster-name/k8s/api/v1/namespaces
> ```

No, I don't think so. AFAIUI, this URL is generated by the kubeapps dashbaord by looking at the [generated k8s API URL](https://github.com/mecampbellsoup/kubeapps/blob/a0164150cd0a0908763618151cdc7e5968d7d1e3/dashboard/src/shared/Kube.ts#L6). So the dashboard's generated URLs will follow the pattern e.g. `/api/clusters/default/can-i` and `/api/clusters/default/namespaces` and not `/k8s/can-i` and `/k8s/namespaces`.

> and want that routed to your `http://gatekeeper-kubernetes-ingress.default` with the extra `/k8s/`, ie. so the path would be rewritten to `/k8s/api/v/namespaces`

Yes, I believe this is what I want!

I think I know what to do now with this NGINX config changeset; I will work on this and ping you once I have pushed something that works. My thinking is that we should add an optional cluster attribute (to the clusters list in `values.yaml`) like `.apiServiceLocation` that, when set, is simply concatenated with the `.apiServiceURL` in the kubeops-related locations, and is added to the rewrite path for what I'm calling the "directly-to-k8s API" location (i.e. `/api/clusters/default`).